### PR TITLE
Java side changes for model import and samediff uptraining

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
java side changes for model import including:
1. Automatic removal of no ops. We need this for automatic differentiation to work properly.
2. Dynamic custom op remove getNumOutputs() which breaks shapes resolution.
3. Adds more type conversions for various fields
4. Adds null check for arrays in SDVariable
5. Better debugging errors for certain op serialization issues
6.  Adds new set shape op
7. Adds new configureFromArguments() for model import handling definitions of configuration objects such as MMulTranspose,Conv2D/3DConfig
8.  Update Flatten and Flatten2D gradients to reflect a proper conversion of the original shape  when going backward
9. Update base reduction custom op to properly add keepDims argument as a boolean.
10. Add missing constructors for splitV for code gen
11. Misc Windows build fixes
12. Updates onnx import uptraining test case for lenet
13.  Fixes reshape bug in c++ when specifying an ndarray for the shape.
14. Fix up reduce test failures with dimensional inputs, big focus around variance/standard deviation related ops including moments, various reduce backpropagation ops 
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ ] Created tests for any significant new code additions.
- [ ] Relevant tests for your changes are passing.

### CODE REVIEW

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/Conv3D.java [hunk 1]
```diff
@@ -273,6 +273,27 @@ public void initFromTensorFlow(NodeDef nodeDef, SameDiff initWith, Map<String, A
         List<SDVariable> inputs = new ArrayList<>();
         inputs.addAll(Arrays.asList(args()));
         inputs.add(f1.get(0));
+
+        if (config == null && !iArguments.isEmpty()) {
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/Conv3D.java:283]: Shouldn't this have the same kind of check as https://github.com/eclipse/deeplearning4j/pull/9501/files#diff-e8b421fce15c6fe13280bf5040a342b65769c2016bbdd5429916c10310b9cdfaR275-R277 ?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/MaxPooling2D.java [hunk 2]
```diff
@@ -142,6 +142,24 @@ public String opName() {
         List<SDVariable> inputs = new ArrayList<>();
         inputs.addAll(Arrays.asList(args()));
         inputs.add(f1.get(0));
+        if(config == null && !iArguments.isEmpty()) {
+            //Perhaps loaded from FlatBuffers - hence we have IArgs but not Config object
+            config = Pooling2DConfig.builder()
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/MaxPooling2D.java]: as above, I guess the same kind of check as in https://github.com/eclipse/deeplearning4j/pull/9501/files#diff-e8b421fce15c6fe13280bf5040a342b65769c2016bbdd5429916c10310b9cdfaR275-R277 would be useful here too.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/Pooling2D.java [hunk 3]
```diff
@@ -144,6 +145,20 @@ public String opName() {
         List<SDVariable> inputs = new ArrayList<>();
         inputs.addAll(Arrays.asList(args()));
         inputs.add(f1.get(0));
+       if(!iArguments.isEmpty() && config == null) {
+           config = Pooling2DConfig.builder()
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/convolution/Pooling2D.java]: as above, I guess the same kind of check as in https://github.com/eclipse/deeplearning4j/pull/9501/files#diff-e8b421fce15c6fe13280bf5040a342b65769c2016bbdd5429916c10310b9cdfaR275-R277 would be useful here too.

**Review by treo — CHANGES_REQUESTED** [review]

**Review by treo — DISMISSED** [review]

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SDVariable.java [hunk 4]
```diff
@@ -64,7 +64,8 @@
 
 
     public SDVariable(@NonNull String varName, @NonNull VariableType varType, @NonNull SameDiff sameDiff, long[] shape, DataType dataType){
-        Preconditions.checkState(dataType != DataType.UNKNOWN, "Unknown datatype is not allowed for SDVariables (variable name: %s)", varName);
+        if(varType != VariableType.PLACEHOLDER)
+            Preconditions.checkState(dataType != DataType.UNKNOWN, "Unknown datatype is not allowed for SDVariables (variable name: %s)", varName);
```
Comment (agibsonccc) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SDVariable.java:69]: @treo this might be a suspect fix for you. Please feel free to take a look. I did this so we could be a bit more flexible with placeholders. Not being able to have an unknown data type for a placeholder didn't make much sense to me.

File: nd4j/samediff-import/samediff-import-api/src/main/kotlin/org/nd4j/samediff/frameworkimport/ImportGraph.kt [hunk 5]
```diff
@@ -218,6 +214,23 @@ open class ImportGraph <GRAPH_TYPE: GeneratedMessageV3,
                             DATA_TYPE, ATTR_DEF_TYPE, ATTR_VALUE_TYPE>): SameDiff {
 
 
+        val newNodeList =  ArrayList<IRNode<NODE_TYPE, TENSOR_TYPE, ATTR_DEF_TYPE, ATTR_VALUE_TYPE, DATA_TYPE>>()
```
Comment (agibsonccc) [inline nd4j/samediff-import/samediff-import-api/src/main/kotlin/org/nd4j/samediff/frameworkimport/ImportGraph.kt]: @treo this is another change that could be suspect...for my test case I ran in to a bunch of problems with :0 as a suffix. You run in to this with tensorflow models as well as tf onnx models.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Reshape.java [hunk 6]
```diff
@@ -59,6 +59,7 @@ public Reshape(SameDiff sameDiff, SDVariable i_v, long[] shape) {
 
     public Reshape(SameDiff sameDiff, SDVariable i_v, SDVariable shape) {
         super(null, sameDiff, new SDVariable[]{i_v, shape});
+        addIArgument(-99);
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Reshape.java]: Can we turn this magic number into a constant?

**Review by treo — CHANGES_REQUESTED** [review]
Only one change request: Please turn the magic numbers for ordering into constants, that way you can put some documentation on the constant.

File: contrib/codegen-tools/codegen/src/main/ops/org/nd4j/codegen/ops/SDBaseOps.kt [hunk 7]
```diff
@@ -898,7 +898,7 @@ fun SDBaseOps() =  Namespace("BaseOps"){
         javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.floating"
         legacy = true
         Input(NUMERIC, "x") { description = "Input variable" }
-        Arg(BOOL, "keepDims") { description = "If true: keep the dimensions that are reduced on (as size 1). False: remove the reduction dimensions"; defaultValue=false }
+        Arg(BOOL, "kemathepDims") { description = "If true: keep the dimensions that are reduced on (as size 1). False: remove the reduction dimensions"; defaultValue=false }
```
Comment (treo) [inline contrib/codegen-tools/codegen/src/main/ops/org/nd4j/codegen/ops/SDBaseOps.kt]: wat?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Reshape.java [hunk 8]
```diff
@@ -68,17 +72,20 @@ public Reshape(INDArray in, long... shape) {
...
         //and a dimension.
-        addIArgument(-99);
+        addIArgument(C_ORDER);
         addIArgument(shape);
     }
 
 
     public Reshape(@NonNull INDArray in, @NonNull INDArray shape, INDArray out) {
         super(null, new INDArray[]{in, shape}, wrapOrNull(out), null, (List<Integer>)null);
+        addIArgument(-99);
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/shape/Reshape.java]: Forgot one here.

**Review by treo — CHANGES_REQUESTED** [review]
You've missed on point with the constants, and apparently you still had focus on your IDE when you started typing about something else in the Ops definition.

**Comment by agibsonccc** [comment]
@quickwritereader there's quite a few changes around reduce operations and dimensions. Could you do a quick scan on the libnd4j bits? Thanks!

**Review by treo — APPROVED** [review]
Didn't look too close at the C++ parts, but otherwise it looks OK now :)

**Review by quickwritereader — APPROVED** [review]
lgtm. But did we decide to keep defined Cygwin there?.
well done.  if it passing tests merge it so I will be able to rebase new refactorings against it

**Comment by agibsonccc** [comment]
@quickwritereader I need to do a part 2 to improve some of the tests yet, but this is a big improvement yes. We did keep cygwin/msys2 setup , but  with more documentation. The main problem was ensuring that we have the right gcc/g++ in the path.
